libpromises/mod_custom.c

```diff
@@ -0,0 +1,651 @@
+            const size_t level_length = equal_sign - level_start;
+            char *const level = xstrndup(level_start, level_length);
+            assert(strlen(level) == level_length);
+
+            JsonElement *log_message = JsonObjectCreate(2);
+            JsonObjectAppendString(log_message, "level", level);
+            JsonObjectAppendString(log_message, "message", message);
+            PromiseModule_LogJson(log_message);
+            JsonArrayAppendObject(log_array, log_message);
+
+            free(level);
+        }
+        else if (line_based)
+        {
+            const char *const equal_sign = strchr(line, '=');
+            assert(equal_sign != NULL);
+            if (equal_sign == NULL)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Promise module sent invalid line: '%s'",
+                    line);
+            }
+            else
+            {
+                const char *const value = equal_sign + 1;
+                const size_t key_length = equal_sign - line;
+                char *const key = xstrndup(line, key_length);
+                assert(strlen(key) == key_length);
+                JsonObjectAppendString(response, key, value);
+                free(key);
+            }
+        }
+        else // JSON protocol:
+        {
+            assert(strlen(line) > 0);
+            assert(response == NULL); // Should be first and only line
+            const char *data = line;  // JsonParse() moves this while parsing
+            JsonParseError err = JsonParse(&data, &response);
+            if (err != JSON_PARSE_OK)
+            {
+                assert(response == NULL);
+                Log(LOG_LEVEL_ERR,
+                    "Promise module '%s' sent invalid JSON",
+                    module->path);
+                free(line);
+                return NULL;
+            }
+        }
+
+        FREE_AND_NULL(line);
+        size = 0;
+    }
+
+    if (line_based)
+    {
+        JsonObjectAppendArray(response, "log", log_array);
+        log_array = NULL;
+    }
+    else
+    {
+        JsonElement *json_log_messages = JsonObjectGet(response, "log");
+
+        // Log messages inside JSON data haven't been printed yet,
+        // do it now:
+        if (json_log_messages != NULL)
+        {
+            size_t length = JsonLength(json_log_messages);
+            for (int i = 0; i < length; ++i)
+            {
+                PromiseModule_LogJson(JsonArrayGet(json_log_messages, i));
+            }
+        }
+
+        JsonElement *merged = NULL;
+        bool had_log_lines = (log_array != NULL && JsonLength(log_array) > 0);
+        if (json_log_messages == NULL && !had_log_lines)
+        {
+            // No log messages at all, no need to add anything to JSON
+        }
+        else if (!had_log_lines)
+        {
+            // No separate log lines before JSON data, leave JSON as is
+        }
+        else if (had_log_lines && json_log_messages == NULL)
```

Comment:
Missing parentheses around the comparison.

Reply:
Fixed.